Support c10::optional<Tensor> in custom C++ autograd function.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37700 Support c10::optional<Tensor> in custom C++ autograd function.**

Certain autograd functions can have optional Tensor arguments. For
this purpose it would be nice to support c10::optional<Tensor> as an argument
for C++ autograd functions.

I've added the appropriate overload to ExtractVariables to ensure this works.
For an example, you can look at D21272807 in terms of how this is used.

Differential Revision: [D21363491](https://our.internmc.facebook.com/intern/diff/D21363491/)

pritamdamania committed:
Support c10::optional<Tensor> in custom C++ autograd function.

Certain autograd functions can have optional Tensor arguments. For
this purpose it would be nice to support c10::optional<Tensor> as an argument
for C++ autograd functions.

I've added the appropriate overload to ExtractVariables to ensure this works.
For an example, you can look at D21272807 in terms of how this is used.

Differential Revision: [D21363491](https://our.internmc.facebook.com/intern/diff/D21363491/)

[ghstack-poisoned]

albanD reviewed (commented):
This looks fine to me.
Maybe add a test that checks that this is working fine?

pritamdamania committed:
Update on "Support c10::optional<Tensor> in custom C++ autograd function."


Certain autograd functions can have optional Tensor arguments. For
this purpose it would be nice to support c10::optional<Tensor> as an argument
for C++ autograd functions.

I've added the appropriate overload to ExtractVariables to ensure this works.
For an example, you can look at D21272807 in terms of how this is used.

Differential Revision: [D21363491](https://our.internmc.facebook.com/intern/diff/D21363491/)

[ghstack-poisoned]

pritamdamania87 commented:
Updated docs and enhanced our `test_custom_operators.py` test for this.

albanD reviewed (approved):
Looks good.
Thanks for the updates.

pritamdamania committed:
Update on "Support c10::optional<Tensor> in custom C++ autograd function."


Certain autograd functions can have optional Tensor arguments. For
this purpose it would be nice to support c10::optional<Tensor> as an argument
for C++ autograd functions.

I've added the appropriate overload to ExtractVariables to ensure this works.
For an example, you can look at D21272807 in terms of how this is used.

Differential Revision: [D21363491](https://our.internmc.facebook.com/intern/diff/D21363491/)

[ghstack-poisoned]
